### PR TITLE
reassign freed pending steps, instead of freed slots

### PR DESCRIPTION
**Pull request by prha**
[INTERNAL_BRANCH=prha/assign_pending_steps]

## Summary & Motivation
Getting into a situation now where an assigned step that is removed before claiming a slot is roadblocking all other pending steps

## How I Tested These Changes
BK

### CODE REVIEW

**Review by alangenfeld — COMMENTED** [review]
did these test changes reproduce the original problem without the fix applied?

**Comment by prha** [comment]
yup.

File: python_modules/dagster/dagster_tests/storage_tests/utils/event_log_storage.py [hunk 1]
```diff
@@ -3976,24 +4012,35 @@ def claim(key, run_id, step_key, priority=0):
...
         assert foo_info.active_run_ids == {run_id}
 
+        assert storage.check_concurrency_claim("foo", run_id, "a").assigned_timestamp is not None
         assert storage.check_concurrency_claim("foo", run_id, "b").assigned_timestamp is not None
+        assert storage.check_concurrency_claim("foo", run_id, "c").assigned_timestamp is None
+        assert storage.check_concurrency_claim("foo", run_id, "d").assigned_timestamp is None
+        assert storage.check_concurrency_claim("foo", run_id, "e").assigned_timestamp is None
+
+        # free the assigned b slot (which has never actually claimed the slot)
+        storage.free_concurrency_slot_for_step(run_id, "b")
```
Comment (prha) [inline python_modules/dagster/dagster_tests/storage_tests/utils/event_log_storage.py:4037]: [1]

**Review by prha — COMMENTED** [review]
[1] this part in particular flexes the error state

**Review by alangenfeld — APPROVED** [review]
in prha we trust

**Pull request merged**